addons/mail/static/src/messaging/utils/deferred/deferred.js

```diff
@@ -0,0 +1,21 @@
+odoo.define('mail.messaging.utils.deferred', function (require) {
+'use strict';
+
+/**
+ * @returns {Deferred}
+ */
+function makeDeferred() {
```

Comment:
It's 100% similar to `makeTestPromise`, but I was afraid to define utility function in `web` and change `makeTestPromise` to return such as promise.

The name "deferred" is totally appropriate, because that's exactly what this is and how it's called by everyone. The sole difference between promise and deferred is the exposure of resolve/reject.

Usage of deferred is deprecated because almost all the time the termination of "asynchronicity" is an output of the asynchronicity itself, and from code maintenance it's easier for programmers to find conditions of resolution/rejection by inspecting a smaller bit of code (i.e. they will eventually find where it's rejected/resolved, without having to look everywhere in their project directory).

Almost of the time... Except in some rare occasions, and timer/throttle here is a good case where deferred is very handy. Without them, business code cannot intercept errors from, resulting, or nested in timers. The reason we need deferred and promise is not sufficient, it's due to how we want to call timer/throttle and how errors could happen:

Let's assumer 2 throttle callers (C1 and C2). C1 is called 1st and C2 is then called. Somehow, the internal of throttle function should reject promise of C1, but as you can see the entry point is not the same, so promises cannot do that... But deferred can!

Reply:
I understand the principle and why we need it, but my main issue is that the resolve/reject is totally leaked to anyone in the outside, including the original the caller.

I think the resolve/reject should instead be enclosed inside the timer, instead of storing the methods on the promise itself, I would prefer to store them on the scope/object that indent to use them, and nowhere else.

Reply:
`resolve`/`reject` functions are directly related to promise, so it makes sense to store on it. If we don't, it's 2 more variables to manage and it makes code of throttle/timer even more complicated to understand.

If you don't want to leak resolve/reject of promise, just don't use deferred in the 1st place. But as I said before, there are rare cases you want them to be leaked, which is precisely what we need in throttle/timer, and deferred make sense there.

Reply:
ok